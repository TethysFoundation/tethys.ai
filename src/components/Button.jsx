import React from 'react';
import PropTypes from 'prop-types';
import { forbidExtraProps } from 'airbnb-prop-types';
import styles from '../assets/stylesheets/button.css';

function getSizeClass(size) {
  let sizeClass = '';

  if (size === 'large') {
    sizeClass = styles.btnLarge;
  } else if (size === 'small') {
    sizeClass = styles.btnSmall;
  }

  return sizeClass;
}

const Button = React.forwardRef((props, ref) => {
  const {
    href, onClick, type, size, newTab, disabled, text,
  } = props;

  const sizeClass = getSizeClass(size);

  if (href) {
    const target = newTab ? '_blank' : null;
    return <a ref={ref} className={[styles.btn, sizeClass].join(' ')} href={href} target={target}>{text}</a>;
  }

  return (
    <button
      ref={ref}
      disabled={disabled}
      type={type}
      className={[styles.btn, sizeClass].join(' ')}
      onClick={onClick}
    >
      {text}
    </button>
  );
});

Button.propTypes = forbidExtraProps({
  href: PropTypes.string,
  onClick: PropTypes.func,
  type: PropTypes.oneOf(['submit']),
  size: PropTypes.oneOf(['small', 'large']),
  disabled: PropTypes.bool,
  newTab: PropTypes.bool,
  text: PropTypes.string.isRequired,
});

Button.defaultProps = {
  href: null,
  onClick: () => {},
  type: null,
  size: 'large',
  disabled: false,
  newTab: false,
};

export default Button;
