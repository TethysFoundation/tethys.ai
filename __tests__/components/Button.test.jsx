import React from 'react';
import { render } from 'react-testing-library';
import Button from '../../src/components/Button';

describe('Button', () => {
  describe('when given an href', () => {
    it('renders an "a" tag', () => {
      const { container } = render(<Button href="/example" text="Submit" />);
      expect(container).toMatchSnapshot();
    });

    it('can render with the small size class', () => {
      const { container } = render(<Button href="/example" text="Submit" size="small" />);
      expect(container).toMatchSnapshot();
    });

    it('adds target _blank when newTab is set to true', () => {
      const { container } = render(<Button href="/example" text="Submit" size="small" newTab />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('when not given an href', () => {
    it('renders a button', () => {
      const { container } = render(<Button text="Submit" />);
      expect(container).toMatchSnapshot();
    });

    it('can render with the small size class', () => {
      const { container } = render(<Button text="Submit" size="small" />);
      expect(container).toMatchSnapshot();
    });
  });
});
