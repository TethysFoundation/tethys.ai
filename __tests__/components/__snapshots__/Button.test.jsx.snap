// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when given an href adds target _blank when newTab is set to true 1`] = `
<div>
  <a
    class="btn btnSmall"
    href="/example"
    target="_blank"
  >
    Submit
  </a>
</div>
`;

exports[`Button when given an href can render with the small size class 1`] = `
<div>
  <a
    class="btn btnSmall"
    href="/example"
  >
    Submit
  </a>
</div>
`;

exports[`Button when given an href renders an "a" tag 1`] = `
<div>
  <a
    class="btn btnLarge"
    href="/example"
  >
    Submit
  </a>
</div>
`;

exports[`Button when not given an href accepts a type prop 1`] = `
<div>
  <button
    class="btn btnSmall"
    type="submit"
  >
    Submit
  </button>
</div>
`;

exports[`Button when not given an href can render with the small size class 1`] = `
<div>
  <button
    class="btn btnSmall"
  >
    Submit
  </button>
</div>
`;

exports[`Button when not given an href renders a button 1`] = `
<div>
  <button
    class="btn btnLarge"
  >
    Submit
  </button>
</div>
`;
